README.md
#Summary

mqtt_gateway is a piece of software that translates between Arduino Manager's (also called AManager) proprietary protocol and MQTT.

#Description

##MQTT

I'll start with ripping off from [the wikipedia article](https://en.wikipedia.org/wiki/MQTT):  "MQTT (formerly Message Queue Telemetry Transport) is a publish-subscribe based "light weight" messaging protocol for use on top of the TCP/IP protocol. It is designed for connections with remote locations where a small code footprint is required and/or network bandwidth is limited."  MQTT is very common, and clients have been written for everything from Arduinos on up to full server-class operating systems.  It's something akin to [SCADA](https://en.wikipedia.org/wiki/SCADA) reinvented for the Internet of Things.  It's use is growing by leaps and bounds, and it's the new standard for passing telemetry for everything from chat clients to burglar alarms to process control to aerospace telemetry.  It's simple enough to run on a 300baud async serial connection to a full IP stack.

##Arduino Manager (AManager)

Arduino Manager is the most wicked cool piece of software you’ve never heard of (website is [here](https://sites.google.com/site/fabboco/home/arduino-manager-for-iphone-ipad)).  It’s a very pretty dashboard that’s completely customizable that talks to an [Arduino](http://www.arduino.cc/) over an IP network.  It basically turns the Arduino into a remote-controlled I/O device.  From the AManager app, you can add virtual switches, graphs, LEDs, meters, etc., and label them as you wish.  Then you assign a pin on the Arduino to each item on the dashboard.  Finally, you click the “generate code” button, and the app spews a bunch of C code that you paste into your Arudino editor, compile, and burn onto your Arduino board.  You’ve now got a virtual control panel or console that can remotely control and read pins on an Arduino board.  There's also a version for the Mac that's quite similar, though considerably buggier.  AManager uses a simple proprietary protocol to communicate between the app and the Arduino.

##The Problem

The problem is that there is nothing remotely like the AManager app for MQTT.  There are some useful tools, many of them quite well implemented, but they're all tools for debugging and doing detailed traces of telemetry data.  There's nothing even remotely like a dashboard like AManager supplies.  Unfortunately, as beautiful and useful is AManager is, it doesn't speak MQTT.  The goal of mqtt_gateway is to marry the two technologies, allowing the use of the AManager app to monitor and control the endless variety of things available via MQTT, and not limited to the twenty I/O pins of a single Arduino device.

#mqtt_gateway

##What Services Does mqtt_gateway Provide?

mqtt_gateway provides translation, throttling, and caching services between MQTT and the proprietary AManager protocols.

###Translation

The AManager protocol is extremely simple.  It consists of ASCII strings containing a variable name, an equals sign, and a value, with a termination hash character carried as the payload of a TCP packet.  For example, the following string contains a variable named "temp" which is set to the value 69.0:

```
temp=69.000#
```

Multiple values can be strung together in a single TCP transaction, like this:

```
temp1=42#temp2=69#
```

Also, the value "1" represents "on" (for switches and LEDs), and "0" represents "off".  That's it, pretty simple.  Values sent from AManager to the Arduino are used to set pins, and values sent from the Arduino to AManager are used to display data (on virtual LEDs, meters, LCD displays, etc).  Normally, a user controls knobs, switches, and sliders, however, a virtual switch, knob, or slider created in AManager can have it's value pre-set (or reset at any time) by setting it's variable name just like any display widget.  This allows for multiple users all with the same switch or button, and it allows the software to pre-set some default control settings, or reset them to the values they were left at when the app was last run.  For example, to pre-set a switch named "light1" to 
"on", the Arduino would send the following string to the app:

```
light1=1#
```

When the AManager app connects to the Arduino after disconnect or app termination, it sends two items after successful TCP connection:  The current time/date (expressed as epoch time) and a list of switches, knobs, and sliders, so that the Arduino has the option of pre-setting those values before the users begins using the interface.  These are sent as follows:

```
$Time=1411603667#
Sync=switch1#Sync=knob2#Sync=slider3#
```

The Arduino may optionally reply by setting the values of each control listed as a "Sync" variable (it may also set them at any arbitrary future time).

MQTT is very free-form, and allows any arbitrary data to be sent on a "topic".  A topic may be thought of as a channel.  Any data sent by a publisher to a topic is received by all subscribers to that topic.  A client may be both a publisher and a subscriber on the same (or any) topics at the same time.  The philosophy of mqtt_gateway is to require all transactions to be on a single topic, and all transactions will be of the form "variable=value".  The most simple service provided by mqtt_gateway is to receive MQTT data, "bundle" it (multiple variables per TCP packet) for efficiency, add the appropriate hash marks, and forward them to the AManager app.  Likewise, it receives variables (possibly "bundled") from the app, and dispenses them one at a time as MQTT data to the specified topic to all subscribers.

###Throttling

AManager is a smartphone app, and as such, it often connects via a cell network.  Cell data is often of limited bandwidth, of high latency (and often high jitter, as well), and rarely inexpensive.  As such, it's prudent to limit the amount of bandwidth used to the minimum amount to satisfy monitoring and control requirements.  MQTT data, on the other hand, can be almost anything at almost any speed.  GPS data might update at 1hz, while temperature measurements might be flowing in 10hz, while RPM measurements are being updated at 100hz.  Rarely is this volume of data needed for human monitoring and control of a process (and if it is, using a $10 iPhone app over a cell network is probably not your best choice).  A way is needed to "smooth out" this data into something usable by humans.  This is solved in mqtt_gateway by comparing the value of each variable received over MQTT with it's previous value.  If they are the same, no data is sent to the AManager app.  Only when a variable changes is the new value sent to the app (with the exception of caching, described next).  For most data, this vastly cuts the amount of data to be transferred.  In experiments done to date, most data is reduced to about 1% of the "full bore" amount, though obviously this varies a great deal based on what kind of data is being sent.  Throttling, in this case, is the logical equivalent of compression, and data with less noise or entropy will "compress" more than "clean and quiet" data.

###Caching

Throttling is good, in terms of controlling the amount of data transferred via cell network, but introduces problems, as well.  If a data point is sent only once every sixty seconds, a user might have to wait up to 59.999 seconds to see that data represented on his AManager screen (assuming that data point is not a duplicate of the previous data point).  Consequently, mqtt_gateway caches values, and sends them at regular intervals, whether or not they've been updated, up to an expiration time.  If the variable has not been updated when that expiration time expires, the values is dropped from the cache.  This prevents long-term false indications of a variable's value (or the existence of that variable).

Also, consider the case where both Alice and Bob are controlling a light bulb.  The both have a switch in their AManager clients labeled "light".  Alice starts her client and turns on the light.  Then Bob starts his client.  Since he missed the event from Alice, he now doesn't know if the light is on or off.  His switch indicates that it's off, but it has never received a value to tell it otherwise.  The only way for Bob to truly know the status of the bulb is to turn it on himself.  Which can cause obvious problems if the bulb shouldn't be on.  mqtt_gateway caches switch, knob, and slider settings indefinitely.  There is no timeout in the cache.  These values are pushed regularly, so that users who have just connected will receive timely status updates on their switch settings.  When a new AManager client is detected (by virtue of the "Sync=x" queries), the switch values are prioritized to be updated as quickly as possible from the cache.

##What Are the Requirements to Run mqtt_gateway?

mqtt_gateway is written in ruby, and has been tested on Linux and OS/X.  It's likely to run as-is (or with few modifications) on Windows.  At present, only one gem is required:

```
sudo gem install mqtt
```

##How Do I Run mqtt_gateway?

You'll need to specify some options on the command line.  At a minimum, you need to specify an MQTT server and a topic:

```
./mqtt_gateway.rb --mqtt mqtt.server.com --topic mydata
```

The rest of the parameters are usually fine using the defaults.

##What Are the Caveats/Problems/Shortcomings of mqtt_gateway?

There are a number of shortcomings of this code (again, it was originally written to solve my own personal use case, not as a full-blown application).  Specifically:

1. The mqtt_gateway will only accept a connection from one AManager app at a time.  This is high on the list of things to change, but for the moment, if you want more than one client, you have to run more than one gateway, and put each on a different TCP port (4444, 4445, 4446, etc).
2. All MQTT traffic is published and subscribed to on a single fixed topic.
3. All traffic on the MQTT side must be of the form "var_name=value".  Simple raw values ("123") won't cause an error, but they won't do anything useful, either.
4. It's probably wise to use a sitter in cron to make sure the client gets restarted if it dies.  It seems robust enough in real use, but there are plenty of potential exceptions that I don't catch, and there is relatively little in the way of error detection.
5. It doesn't always correctly detect the disconnect of AManager immediately.  It often takes 15-30 seconds before it realizes the client is gone and is ready to accept a new connection.
6. There is currently no provision for authentication to the MQTT server.  This will change, but the current version of code requires an open MQTT server (mosquitto works great).
7. There's something going on I don't understand which often corrupts $rx when the client disconnects. As a result, $rx is generated fresh for each client, but is updated from the previous YAML save, if that's available.

##Where Do I Get It?

If you're reading this somewhere other than github, click [here](https://github.com/jfrancis42/amanager-mqtt-gateway) to visit the github page for dowload.

##I Have an Idea!

That's great, and I'm open to hearing it. Just keep in mind this is a little personal project that I'm giving away for those who might want to play with it (or base something else on it), it's not a full on industrial-strength open source project. I have a day job, a family, and many other hobbies and projects, so ideas with code attached will get looked at (and/or implemented) much faster than those without.


Jeff Francis


NEWS
Sep 25, 2014:  Everything works, no known serious bugs.


mqtt_gateway.rb
#!/usr/bin/ruby
# -*- coding: utf-8 -*-

# Jeff Francis, N0GQ, jeff@gritch.org
# http://www.gritch.org/amanager/
#
# mqtt_gateway is a piece of software that translates between Arduino
# Manager's (also called AManager) proprietary protocol and MQTT.
#
# 25.September.2014
#
# You'll need two gems that aren't part of the default ruby distro to
# use this software: mqtt and trollop.
#
# On Linux/BSD/OSX boxes, you can install them like this:
#
# sudo gem install mqtt
# sudo gem install trollop

# This sets up all of our default options.
DEBUG=false
QUEUE=false
max_at_once=6
max_old_send=3
discard_age=300
update_age=30
$port=4444
$stash_file=nil

# We need these gems to work.  These are built in...
require 'socket'
require 'time'
require 'yaml'

# ...and these are not.
require 'mqtt'
require 'trollop'

# Process the options.
opts=Trollop::options do
  opt :mqtt, "MQTT Server", :type => :string
  opt :topic, "MQTT Topic", :type => :string
  opt :port, "TCP port to listen for AManager connections (default 4444)", :type => :string
  opt :batch, "Maximum variable updates to AManager per transaction (default 6)", :type => :string
  opt :old, "Maximum old variables to refresh per transaction (default 3)", :type => :string
  opt :update, "Maximum age in seconds to resend old variable (default 30)", :type => :string
  opt :discard, "Maximum age in seconds to discard non-updated variables (default 300)", :type => :string
  opt :yaml, "Cache current status in a YAML file", :type => :string
  opt :debug, "Show extra debug info (not useful for most users)"
  opt :queue, "Dump the queue to STDOUT every few seconds (very verbose)"
end

# Override default discard time.
if opts[:discard_given]
  discard_age=opts[:discard].to_i
end

# Override default update time.
if opts[:update_given]
  update_age=opts[:update].to_i
end

# Override max number of old messages sent.
if opts[:old_given]
  max_old_send=opts[:old].to_i
end

# Override default max number of variables per transaction.
if opts[:batch_given]
  max_at_once=opts[:batch].to_i
end

# Turn on debug mode.
if opts[:debug_given]
  DEBUG=TRUE
end

# Turn on super-debug mode (queue dumping).
if opts[:queue_given]
  QUEUE=TRUE
end

# Specify the name of the mqtt server.
if opts[:mqtt_given]
  $server=opts[:mqtt]
else
  puts "--mqtt is mandatory"
  exit
end

# Specify the name of the topic.
if opts[:topic_given]
  $topic=opts[:topic]
else
  puts "--topic is mandatory"
  exit
end

# Specify the port to listen on.
if opts[:port_given]
  $port=opts[:port].to_i
end

# Specify the name of the YAML dump file.
if opts[:yaml_given]
  $stash_file=opts[:yaml]
end

$rxmutex=Mutex.new
$rx=nil
$send=Array.new

# Serialize an object as YAML.
def write_object_as_yaml(filename, object)
  File.open(filename, 'w') do |io|
    YAML::dump(object, io)
  end
end

# Deserialize a YAML object.
def read_object_from_yaml(filename, default={})
  return default unless File.exists? filename
  File.open(filename) do |io|
    return YAML::load(io)
  end
end

# This class holds all the MQTT variable info.
class Thing
  attr_accessor :name, :value, :rtime, :stime, :updated, :phoneset

  def initialize(name, value, rtime)
    @name = name.to_s
    @value = value
    @rtime = rtime
    @stime = 0
    @updated = true
    @phoneset = false
  end

  def to_s
    "Name: #{@name}\nValue: #{@value}\nReceived: #{@rtime.to_s}\nSent: #{@stime.to_s}\nUpdated: #{@updated}\nPhone-Set: #{@phoneset}"
  end

  def to_s_short
    "#{@name} #{@value} #{@rtime.to_i} #{@stime.to_i} #{@updated} #{@phoneset}"
  end
end

# This hash holds all of the Thing objects.
data=Hash.new

# Recalculate and update the queue depth every n seconds.
def calc_queue(n)
  depth=0
  # Wait one "cycle" to start.
  sleep n
  loop {
    $rxmutex.synchronize {
      depth=0
      $rx.each_key do |key|
        if $rx[key].updated
          depth=depth+1
        end
      end
    }
    if !$rx.has_key?('Q')
      $rx['Q']=Thing.new('Q',0,Time.now())
    end
    $rx['Q'].value=depth
    $rx['Q'].updated=true
    puts "Queue:  #{depth}" if DEBUG
    if QUEUE
      $rx.each_key do |key|
        puts "#{key} #{$rx[key].to_s_short}"
      end
    end
    # Stash our current state.
    if $stash_file
      write_object_as_yaml($stash_file, $rx)
    end
    sleep n
  }
end

# Subscribe to the given channel and update the data structure as new
# values come in.
def mqtt_sub(myserver, mytopic)
  puts "mqtt_sub() thread started..." if DEBUG
  MQTT::Client.connect(myserver) do |mqtt|
    puts "mqtt_sub() thread connected..." if DEBUG
    mqtt.get(mytopic) do |topic,rxmsg|
      now=Time.now()
      this,that=rxmsg.split('=')
      $rxmutex.synchronize {
        if $rx.has_key?(this)
          $rx[this].rtime=now
          if $rx[this].value==that
            puts "Redundant value: #{rxmsg}" if DEBUG
          else
            puts "New value: #{rxmsg}" if DEBUG
            $rx[this].updated=true
            $rx[this].value=that
          end
        else
          puts "New variable: #{rxmsg}" if DEBUG
          $rx[this]=Thing.new(this, that, now)
        end
      }
    end
  end
end

# Start the queue depth thread.
qthread=Thread.new { calc_queue(5) }
qthread.abort_on_exception=true

# Start the MQTT subscriber thread.
subthread=Thread.new { mqtt_sub($server, $topic) }
subthread.abort_on_exception=true

# And here we go with the main thread. Run forever. Well, or until
# something breaks...
loop {
  $rx=Hash.new
  if $stash_file
    $rx=read_object_from_yaml($stash_file) if File.exist?($stash_file)
  end

  # Listen for AManager to connect, then start processing data.
  a=TCPServer.new(nil,$port)
  connection=a.accept
  client_connected=true

  # This is for sending stuff from the phone to MQTT.
  MQTT::Client.connect($server) do |c|

    # Loop forever (or at least until the app closes/sleeps).
    while client_connected do
      # Is there data waiting for us (from the phone)?
      ready=IO.select([connection],nil,nil,1)

      # Process stuff sent by the phone.  Just sent it as it comes, no
      # need for the cool buffering like on $rx.
      if ready
        r=connection.recv(1024).split('#')
        if r.length>0
          puts "phone sent: " + r.join('#') + '#' if DEBUG
          r.each do |n|
            (name,value)=n.split('=')
            if name=="Sync"
              # This is special. "Wake up" any values asked for by the
              # phone (they're syncing state, so the sooner we send
              # it, the sooner the switches on the phone will sync up
              # and show the right values). We set the sent time to
              # zero, the received time to now, and updated to true to
              # ensure that these values go to the top of the list of
              # things to send.
              puts "sync: #{name}=#{value}" if DEBUG
              if $rx.has_key?(value)
                $rxmutex.synchronize {
                  $rx[value].rtime=Time.now()
                  $rx[value].stime=0
                  $rx[value].updated=true
                }
              end
            else
              # If the user sends data from the phone, throw it in the
              # queue at high priority for sending. Also, note that
              # it's from the phone (this makes it immune from
              # purging).
              puts "sending: #{n}" if DEBUG
              tmp_msg="#{name}=#{value}"
              c.publish($topic, tmp_msg)
              if $rx.has_key?(name)
                $rxmutex.synchronize {
                  $rx[name].rtime=Time.now()
                  $rx[name].stime=0
                  $rx[name].value=value
                  $rx[name].updated=true
                  $rx[name].phoneset=true
                }
              else
                $rxmutex.synchronize {
                  $rx[name]=Thing.new(name,value,Time.now())
                }
              end
            end
          end
        end
      end
      
      # Loop through our Thing data structure and send what needs sending.
      send_msg=""
      $rxmutex.synchronize {
        
        now=Time.now()

        # First, throw away stuff that's really old (the value hasn't
        # been updated in discard_age seconds). Values sent from the
        # phone are immune from discarding.
        $rx.each_key do |key|
          if now.to_i-$rx[key].rtime.to_i>discard_age and !$rx[key].phoneset
            $rx.delete(key)
          end
        end

        # Now make a list of data that's newer than discard_age, but
        # hasn't been sent to the phone in at least update_age seconds,
        # but doesn't contain an updated value (ie, a list of stuff to
        # re-send the old values to the phone for to keep the phone screen
        # updated).
        old=Hash.new
        $rx.each_key do |key|
          if now.to_i-$rx[key].stime.to_i>update_age
            old[key]=true
          end
        end
        puts "#{old.length} old values to send" if DEBUG

        # Finally, make a list if data that's got new values (the stuff
        # that's changed since the last time we sent it).  These values
        # get priority in the sending queue.
        updated=Hash.new
        $rx.each_key do |key|
          if $rx[key].updated
            updated[key]=true
          end
        end
        puts "#{updated.length} updated values to send" if DEBUG
        
        # Now figure out what we're actually going to send to the phone
        # and send it.  This gets complicated.  We can send a maximum of
        # max_at_once values per update to the phone.  If the sum of old
        # and updated values to be sent is less than or equal to this, we
        # send everything.  Otherwise, data gets sent based on space
        # available, with priority going to new data (but old data always
        # gets up to max_old_send slots).
        send_count=0
        send_msg=""
        oldcount=old.length
        updatedcount=updated.length
        
        # Easy case first.  If the total count <= the maximum allowed,
        # just send everything.
        if oldcount+updatedcount<=max_at_once
          updated.each_key do |key|
            send_msg=send_msg+key+"="+$rx[key].value.to_s+"#"
            $rx[key].stime=now
            $rx[key].updated=false
            send_count=send_count+1
          end
          old.each_key do |key|
            send_msg=send_msg+key+"="+$rx[key].value.to_s+"#"
            $rx[key].stime=now
            $rx[key].updated=false
            send_count=send_count+1
          end
        else
          # If that's not the case, break it down and send some of each.

          # First, send up to max_old_send old messages.
          if oldcount>0
            old.each_key do |key|
              if send_count<max_old_send
                send_msg=send_msg+key+"="+$rx[key].value.to_s+"#"
                $rx[key].stime=now
                $rx[key].updated=false
                send_count=send_count+1
                oldcount=oldcount-1
                old.delete(key)
              end
            end
          end

          # Now add new messages until we either run out of them, or reach
          # maximum message length.
          if updatedcount>0
            updated.each_key do |key|
              if send_count<max_at_once
                send_msg=send_msg+key+"="+$rx[key].value.to_s+"#"
                $rx[key].stime=now
                $rx[key].updated=false
                send_count=send_count+1
              end
            end
          end

          # Finally, if there's any space left, send more old messages (if
          # there are any).
          if oldcount>0 and send_count<max_at_once
            old.each_key do |key|
              if send_count<max_at_once
                send_msg=send_msg+key+"="+$rx[key].value.to_s+"#"
                $rx[key].stime=now
                $rx[key].updated=false
                send_count=send_count+1
                old.delete(key)
              end
            end
          end

        end
      }

      # Send the data to the phone (if any).
      if send_msg.length>0
        puts "->#{send_msg}<-" if DEBUG
        begin
          connection.write(send_msg)
        rescue Errno::EPIPE
          puts "Client disconnected..." if DEBUG

          # Close open stuff.
          connection.close
          a.close

          # Write the latest data to the file.
          if $stash_file
            write_object_as_yaml($stash_file, $rx)
          end

          # Tell the loop to start over.
          client_connected=false
        end
      end
    end
  end
}
